userland/libtock/alarm_timer.c

```diff
@@ -10,162 +10,115 @@ static int cmp_exp(uint32_t now, uint32_t exp0, uint32_t exp1) {
   return (exp0 - now) - (exp1 - now);
 }
 
-struct alarm {
-  uint32_t t0;
-  uint32_t expiration;
-  subscribe_cb *callback;
-  void* ud;
-};
-
-typedef struct {
-  alarm_t** data;
-  int capacity;
-  int size;
-} heap_t;
-
-static heap_t alarm_heap = {
-  .data     = NULL,
-  .capacity = 0,
-  .size     = 0,
-};
-
-static int heap_insert(alarm_t* alarm) {
-  if (alarm_heap.capacity - alarm_heap.size <= 0) {
-    // Heap too small! Make it bigger
-    int new_capacity = (alarm_heap.capacity + 1) * 2;
-    alarm_heap.data = (alarm_t**)realloc(alarm_heap.data,
-                                         new_capacity * sizeof(alarm_t*));
-    if (alarm_heap.data == NULL) {
-      return TOCK_ENOMEM;
-    }
-    alarm_heap.capacity = new_capacity;
+static alarm_t* root = NULL;
+
+static void root_insert(alarm_t* alarm) {
+  if (root == NULL) {
+    root       = alarm;
+    root->next = NULL;
+    root->prev = NULL;
+    return;
   }
 
-  // insert it at the end...
-  int idx = alarm_heap.size;
-  alarm_heap.data[idx] = alarm;
-  alarm_heap.size++;
-
-  // then up-heap...
-  while (idx != 0) {
-    int parent_idx  = (idx - 1) / 2;
-    alarm_t *parent = alarm_heap.data[parent_idx];
-    if (cmp_exp(alarm->t0, alarm->expiration, parent->expiration) < 0) {
-      alarm_heap.data[idx]        = parent;
-      alarm_heap.data[parent_idx] = alarm;
-      idx = parent_idx;
-    } else {
-      break;
+  alarm_t **cur = &root;
+  alarm_t *prev = NULL;
+  while (*cur != NULL) {
+    if (cmp_exp(alarm->t0, alarm->expiration, (*cur)->expiration) < 0) {
+      // insert before
+      alarm_t *tmp = *cur;
+      *cur        = alarm;
+      alarm->next = tmp;
+      alarm->prev = prev;
+      tmp->prev   = alarm;
+      return;
     }
+    prev = *cur;
+    cur  = &prev->next;
   }
+  // didn't return, so prev points to the last in the list
+  prev->next  = alarm;
+  alarm->prev = prev;
+  alarm->next = NULL;
 
-  return 0;
 }
 
-static alarm_t* heap_pop(uint32_t now) {
-  if (alarm_heap.size == 0) {
+static alarm_t* root_pop(void) {
+  if (root == NULL) {
     return NULL;
-  }
-
-  alarm_t* ret = alarm_heap.data[0];
-
-  // swap leaf element to root
-  alarm_heap.size--;
-  if (alarm_heap.size == 0) {
-    return ret;
-  }
-  alarm_t *alarm = alarm_heap.data[alarm_heap.size];
-  alarm_heap.data[0] = alarm;
-
-  // sift-down
-  int idx = 0;
-  while (idx < alarm_heap.size) {
-    int childl_idx = (idx + 1) * 2;
-    int childr_idx = (idx + 2) * 2;
```

Comment:
ugh... you're probably right... and it would only manifest when there are more than 2 outstanding timers i think, which would explain why it showed up under a particular workload...